bundle/edu.gemini.lchquery.servlet/src/main/scala/edu/gemini/lchquery/servlet/LchQueryFunctor.scala

```diff
@@ -0,0 +1,230 @@
+import edu.gemini.pot.spdb.{DBAbstractQueryFunctor, IDBDatabaseService, IDBFunctor, IDBParallelFunctor}
+import edu.gemini.skycalc.{DDMMSS, HHMMSS}
+import edu.gemini.spModel.gemini.obscomp.SPSiteQuality
+import edu.gemini.spModel.obs.{ObsClassService, ObservationStatus}
+import edu.gemini.spModel.rich.shared.immutable._
+import edu.gemini.spModel.target.SPTarget
+import edu.gemini.spModel.target.env.TargetEnvironment
+import edu.gemini.spModel.too.Too
+
+import scala.collection.JavaConverters._
+import scala.util.{Failure, Success, Try}
+
+
+class LchQueryFunctor(queryType: LchQueryFunctor.QueryType,
+                      programParams: List[(LchQueryParam[ISPProgram], String)],
+                      observationParams: List[(LchQueryParam[ISPObservation], String)])
+  extends DBAbstractQueryFunctor with IDBParallelFunctor {
+
+  // The current QueryResult. Must be mutable.
+  var queryResult: QueryResult = new QueryResult() {
+    setProgramsNode(new ProgramsNode())
+  }
+
+  // Called once per program by IDBQueryRunner implementation.
+  override def execute(db: IDBDatabaseService, node: ISPNode, principals: util.Set[Principal]): Unit = {
+    val prog = node.asInstanceOf[ISPProgram]
+
+    def programMatches: Boolean = programParams.forall {
+      case (LchQueryParam(_, valueMatcher), paramValue) => valueMatcher.matches(paramValue, prog)
+    }
+    def observationMatches(obs: ISPObservation): Boolean = observationParams.forall {
+      case (LchQueryParam(_, valueMatcher), paramValue) => valueMatcher.matches(paramValue, obs)
+    }
+
+    Try {
+      // See if an ISPProgram matches the query specifications.
+      if (programMatches) {
+        val matchingObs = prog.getAllObservations.asScala.toList.filter(observationMatches)
+        if (matchingObs.nonEmpty) {
+          addProgram(prog, if (queryType == LchQueryFunctor.QueryType.ProgramQuery) Nil else matchingObs)
+        }
+      }
+    } match {
+      case Success(_) =>
+      case Failure(t) =>
+        LchQueryFunctor.Log.log(Level.SEVERE, "problem running LchQueryFunctor", t)
+        throw new RuntimeException(t)
+    }
+  }
+
+
+  private def addProgram(prog: ISPProgram, obsList: List[ISPObservation]): Unit = {
+    import LchQueryParam.{ISPProgramExtractors, ISPObservationExtractors, ToYesNo}
+
+    def makeTargetNode(target: SPTarget, env: TargetEnvironment): Option[Serializable] = {
+      def targetType: Option[String] = {
+        if (env.isBasePosition(target)) Some("Base")
+        else if (env.isUserPosition(target)) Some("User")
+        else if (env.isGuidePosition(target)) {
+          for {
+            gg  <- env.getGroups.asScala
+            gpt <- gg.getAll.asScala
+            if gpt.containsTarget(target)
+          } yield gpt.getGuider.getKey
+        }.headOption
+        else None
+      }
+
+      if (target.isNonSidereal) {
+        Some(new NonSidereal() {
+          setName(target.getName)
+          setType(targetType.orNull)
+        })
```

Comment:
It took me a minute to figure out but I guess we're hamstrung by the the ODB browser library supplied and used by the LCH?  That is `NonSidereal` and `Sidereal` comes from that and they require a `String` target type, etc.